StRoot/StFcsWaveformFitMaker/StFcsWaveformFitMaker.cxx

```diff
@@ -151,51 +238,115 @@ int StFcsWaveformFitMaker::InitRun(int runNumber) {
         LOG_ERROR << "StFcsWaveformFitMaker initializing failed due to no StFcsDb" << endm;
         return kStErr;
     }
+    mDbPulse = static_cast<StFcsDbPulse*>(GetDataSet("fcsPulse"));
+    if (!mDbPulse) {
+        LOG_ERROR << "StFcsWaveformFitMaker initializing failed due to no StFcsDbPulse" << endm;
+        return kStErr;
+    }
+    mDbPulse->setTail(mTail);
+    if( mPulseFit==0 ){ mPulseFit = new StFcsPulseAna(); SetupDavidFitterMay2022(); mPulseFit->setDbPulse(mDbPulse); }
     
     if(mFilename){
       mCanvas=new TCanvas("FCSWaveFormFit","FCSWaveFormFit",10,10,2000,2000);
       gStyle->SetOptStat(0);
+      char file[100];
+      if( mFitDrawOn>=0 ){
+	sprintf(file,"%s.pdf[",mFilename);//Opens fileA but does not save anything
+	mCanvas->Print(file);
+      }
     }
     if(mMeasureTime){
-      mTime=new TH1F("FitTime","FitTime; msec",200,0,6);
+      mTime=new TH1F("FitTime","FitTime;msec",1000,0,1000);
```

Comment:
I am using this line to comment on a common problem which this PR seems to introduce in quite a few other places.
The mTime histogram can be allocated multiple times during a job when the framework calls InitRun(), however, the corresponding object is deleted only once in the destructor, hence causing a memory leak. Depending on the logic you are trying to implement here, you may need to allocate histograms only once in the Init() method. It may be also easier to rely on the automatic object lifetime than trying to manually manage the memory like you do.

Reply:
I thought InitRun was only called once during a job. that is why I put it there. I agree that I only need these histograms created once per job. Where should I move the histogram allocations so that they only get created once?

Reply:
Move to Init(). In principle, InitRun can be called multiple times per job if you are reading file(s) from multiple runs.

Reply:
If it needs to be done only once, is there a reason not to initialize the histograms in the constructor? Also, if there is no good reason to allocate them with the `new` operator, you can avoid the many `delete` statements in the destructor.

Reply:
NB: During StMaker::Init(), any newly instantiated histograms are added to the maker's "hist" subdirectory (in TDataset terms) which allows them to be found using the TDataset methods and enables inclusion in the BFC's hist.root output files.

//              Add the Maker histograms in the Maker histograms list
//              and remove it from the ROOT system directory
      gROOT->cd();
      TIter nextHist(gDirectory->GetList());
      Int_t ready = !objLast;
      while((objHist=nextHist())) {// loop over gDirectory
        if (!ready && objHist!=objLast)         continue;
        ready = 1999;
        if (objHist==objLast)                   continue;
        if (!objHist->InheritsFrom("TH1"))      continue;

//              Move the histogram from the ROOT list into the "maker's" list
        ((TH1*)objHist)->SetDirectory(0);
        maker->AddHist((TH1*)objHist);
      }

Reply:
Moving them to the constructor means that these histograms will always be created, whereas now they are only created when a certain test level is set using *setTest()*. I thought this would be best since in the production these histograms don't need to be created as they are mostly used to quantify how well the algorithm is working. If I move them to the constructor then I would have to add an argument for the test level and file name where those histograms will be saved. I think doing that will only make this class more complicated so I left them in Init(). That way whoever calls this class can turn testing on or off before it hits the chain.